store/tracekvstore.go

```
@@ -0,0 +1,128 @@
+func (tkv *TraceKVStore) Delete(key []byte) {
+	tkv.writeOperation(writeOp, key, nil)
+	tkv.parent.Delete(key)
+}
+
+// Has implements the KVStore interface. It delegates the Has call to the
+// parent KVStore.
+func (tkv *TraceKVStore) Has(key []byte) bool {
+	return tkv.parent.Has(key)
+}
+
+// Prefix implements the KVStore interface.
+func (tkv *TraceKVStore) Prefix(prefix []byte) KVStore {
+	return prefixStore{tkv, prefix}
+}
+
+// Iterator implements the KVStore interface. It delegates the Iterator call
+// the to the parent KVStore.
+func (tkv *TraceKVStore) Iterator(start, end []byte) sdk.Iterator {
+	return tkv.iterator(start, end, true)
+}
+
+// ReverseIterator implements the KVStore interface. It delegates the
+// ReverseIterator call the to the parent KVStore.
+func (tkv *TraceKVStore) ReverseIterator(start, end []byte) sdk.Iterator {
+	return tkv.iterator(start, end, false)
+}
+
+// iterator facilitates iteration over a KVStore. It delegates the necessary
+// calls to it's parent KVStore.
+func (tkv *TraceKVStore) iterator(start, end []byte, ascending bool) sdk.Iterator {
+	if ascending {
+		return tkv.parent.Iterator(start, end)
+	}
+
+	return tkv.parent.ReverseIterator(start, end)
+}
+
+// GetStoreType implements the KVStore interface. It returns the underlying
+// KVStore type.
+func (tkv *TraceKVStore) GetStoreType() sdk.StoreType {
+	return tkv.parent.GetStoreType()
+}
+
+// CacheWrap implements the KVStore interface. It panics as a TraceKVStore
+// cannot be cache wrapped.
+func (tkv *TraceKVStore) CacheWrap() sdk.CacheWrap {
+	panic("you cannot CacheWrap a TraceKVStore")
+}
+
+// writeOperation writes a KVStore operation to the underlying io.Writer as
+// JSON-encoded data.
+func (tkv *TraceKVStore) writeOperation(op operation, key, value []byte) {
+	raw, err := json.Marshal(traceOperation{
+		Timestamp: time.Now().UTC().Nanosecond(),
+		Operation: op,
+		Key:       string(key),
+		Value:     string(value),
+	})
+
+	if err == nil {
```

Does this fail sometimes? Might be better to panic.

---

Mhmmm, I thought about that. It should never panic really. Do we really wanna panic on a failure to trace some operation? It would make debugging tricky maybe, but idk if it warrants a panic.

---

I think we should; better than to fail silently (maybe the machine ran out of RAM? Hard to imagine a case when this would panic). This is an opt-in setting anyways; most users won't use a `TraceKVStore`.

---

Fair point 👍